Typescript/en-us/Lesson03/AlienFinder.ts

```diff
@@ -0,0 +1,293 @@
+                break;
+            }
+            case 2: {
+                let left: number = a.xna / 1E4;
+                let right: number = a.xna % 1E2;
+                a.xna = left * 1E4 + randomDigit * 1E2 + right;
+                this.aliens.set(id, a);
+                break;
+            }
+            case 3: {
+                let left: number = a.xna / 1E2;
+                a.xna = left * 1E2 + randomDigit;
+                this.aliens.set(id, a);
+                break;
+            }
+            default: {
+                break;
+            }
+        }
+    }
+
+    // event selection
+    @constant
+    public d6(): number {
+        return this.randomNumber(Blockchain.currentHeight) % 6;
+    }
+
+    // xna promotion
+    @constant
+    public d10(): number {
+        return this.randomNumber(Blockchain.currentHeight) % 10;
+    }
+
+    // event judgement
+    @constant
+    public d100(): number {
+        return this.randomNumber(Blockchain.currentHeight) % 100;
+    }
+
+    // enemy selection
+    @constant
+    public dN(n: number): number {
+        return this.randomNumber(Blockchain.currentHeight) % n + 1;
+    }
+
+    private reward(alien: Alien) {
+        let attribute: number = this.d6() / 2;
+        let value: number = this.d10();
+
+        switch(attribute) {
+            case 0: {
+                alien.xna += value * 1E4;
+                notifyReward('strength', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            case 1: {
+                alien.xna += value * 1E2;
+                notifyReward('speed', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            case 2: {
+                alien.xna += value;
+                notifyReward('weight', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            default: {
+                break;
+            }
+        }
+    }
+
+    private punish(alien: Alien) {
+        let attribute: number = this.d6() / 2;
+        let value: number = this.d10();
+
+        switch(attribute) {
+            case 0: {
+                alien.xna -= this.d10() * 1E4;
+                notifyPunishment('strength', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            case 1: {
+                alien.xna -= this.d10() * 1E2;
+                notifyPunishment('speed', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            case 2: {
+                alien.xna -= this.d10();
+                notifyPunishment('weight', value);
+                this.aliens.set(alien.id, alien);
+                break;
+            }
+            default: {
+                break;
+            }
+        }
+    }
+
+    private judge(alien: Alien, value: number) {
+        if (value + this.d100() < 200) {
```

Comment:
`value` = 0-99, and `this.d100()` = 0-99; so this would be always true, right?

Reply:
This event should reward low `weight` in the weight event, and reward high `speed` and `strength` in their events. 

If we use `value + this.d100() < 100`, it rewards low `value`. So for speed and strength event, maybe we can pass in `99 - speed` as the parameter?

Reply:
> If value = 0-99, and this.d100() = 0-99;
> then this boolean check would be always true, right?

Well. it must be a bug